src/chainstate/coordinator/mod.rs

```diff
@@ -0,0 +1,363 @@
+                    if let Err(e) = inst.process_ready_blocks() {
+                        warn!("Error processing new stacks block: {:?}", e);
+                    }
+                },
+                i if i == event_burn_block => {
+                    // pop operation off of receiver
+                    debug!("Received new burn block notice");
+                    ready_oper.recv(&burn_block_channel).unwrap();
+                    if let Err(e) = inst.handle_new_burnchain_block() {
+                        warn!("Error processing new burn block: {:?}", e);
+                    }
+                },
+                i if i == event_stop => {
+                    CoordinatorCommunication::cleanup_singleton();
+                    return
+                },
+                _ => {
+                    unreachable!("Ready channel for non-registered channel");
+                },
+            }
+        }
+    }
+}
+
+impl <'a, T: BlockEventDispatcher> ChainsCoordinator <'a, T, ()> {
+    #[cfg(test)]
+    pub fn test_new(burnchain: &Burnchain, path: &str) -> ChainsCoordinator<'a, T, ()> {
+        let burnchain = burnchain.clone();
+
+        let sortition_db = SortitionDB::open(&burnchain.get_db_path(), true).unwrap();
+        let burnchain_blocks_db = BurnchainDB::open(&burnchain.get_burnchaindb_path(), false).unwrap();
+        let chain_state_db = StacksChainState::open(false, 0xdeadbeef, &format!("{}/chainstate/", path)).unwrap();
+
+        let canonical_sortition_tip = SortitionDB::get_canonical_sortition_tip(sortition_db.conn()).unwrap();
+
+        ChainsCoordinator {
+            canonical_chain_tip: None,
+            canonical_sortition_tip: Some(canonical_sortition_tip),
+            canonical_pox_id: None,
+            burnchain_blocks_db,
+            chain_state_db,
+            sortition_db,
+            burnchain,
+            dispatcher: None,
+            notifier: ()
+        }
+    }
+}
+
+impl <'a, T: BlockEventDispatcher, N: CoordinatorNotices> ChainsCoordinator <'a, T, N> {
+    pub fn handle_new_burnchain_block(&mut self) -> Result<(), Error> {
+        // Retrieve canonical burnchain chain tip from the BurnchainBlocksDB
+        let canonical_burnchain_tip = self.burnchain_blocks_db.get_canonical_chain_tip()?;
+
+        // Retrieve canonical pox id (<=> reward cycle id)
+        let mut canonical_sortition_tip = self.canonical_sortition_tip.clone()
+            .expect("FAIL: no canonical sortition tip");
+
+        // Retrieve all the direct ancestors of this block with an unprocessed sortition 
+        let mut cursor = canonical_burnchain_tip.block_hash.clone();
+        let mut sortitions_to_process = VecDeque::new();
+
+        // We halt the ancestry research as soon as we find a processed parent
+        while !(self.sortition_db.is_sortition_processed(&cursor, &canonical_sortition_tip)?) {
+            let current_block = self.burnchain_blocks_db.get_burnchain_block(&cursor)
```

Comment:
This method usage seems broken, and it breaks this entire method as a result.  When we download and store burnchain transactions, we _filter out transactions with invalid consensus hashes_.  Meaning, when we go and re-process a burnchain block due to a PoX fork getting invalidated, the burnchain DB will _NOT_ have the burnchain transactions we want.

One of two things needs to happen:
* We need to store all burnchain transactions that _might_ be Stacks transactions, if it weren't for the consensus hash.  This means doing a two-pass burnchain transaction validation:  one pass to check everything but the consensus hash, and another pass to check the consensus hash when the transaction is getting re-considered here.
* We need to re-download the burnchain blocks on PoX fork reprocessing.

Reply:
> We need to store all burnchain transactions that might be Stacks transactions, if it weren't for the consensus hash. This means doing a two-pass burnchain transaction validation: one pass to check everything but the consensus hash, and another pass to check the consensus hash when the transaction is getting re-considered here.

Yes -- that's what's currently happening -- the BurnchainDB stores all the burnchain transactions that might be Stacks transactions. The `validate()` functions are only called when the SortitionDB processes a sortition and the BlockSnapshot is created.

Reply:
I don't see where this happens in this PR?  Was this done in an earlier PR?

Reply:
Yeah -- that happens in `burnchains/db.rs:store_new_burnchain_block` (invoked by `process_blocks` in burnchain.rs). That was done in the SortitionDB/BurnchainDB PR (#1725).

Reply:
I see. Thanks!